File: app/src/main/java/com/example/zhpan/circleviewpager/view/DrawableIndicator.java
package com.example.zhpan.circleviewpager.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.AttributeSet;

import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatDrawableManager;
import androidx.core.graphics.drawable.DrawableCompat;

import com.zhpan.indicatorview.base.BaseIndicatorView;


/**
 * @ author : zhouweibin
 * @ time: 2019/12/18 17:04.
 * @ desc: 选中与未选中的图片长宽可能不一样
 **/
public class DrawableIndicator extends BaseIndicatorView {
    // 选中与未选中的图片
    private Bitmap mCheckedBitmap, mNormalBitmap;
    // 图片之间的间距
    private int mIndicatorPadding;
    // 选中图片的宽高
    private int mCheckedBitmapWidth, mCheckedBitmapHeight;
    //未选中图片的宽高
    private int mNormalBitmapWidth, mNormalBitmapHeight;
    private IndicatorSize mIndicatorSize;
    private boolean normalCanResize = true;
    private boolean checkCanResize=true;

    public DrawableIndicator(Context context) {
        this(context, null);
    }

    public DrawableIndicator(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DrawableIndicator(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int maxHeight = Math.max(mCheckedBitmapHeight, mNormalBitmapHeight);
        int realWidth = mCheckedBitmapWidth + (mNormalBitmapWidth + mIndicatorPadding) * (getPageSize() - 1);
        setMeasuredDimension(realWidth, maxHeight);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (getPageSize() > 1 && mCheckedBitmap != null && mNormalBitmap != null) {
            for (int i = 1; i < getPageSize() + 1; i++) {
                int left;
                int top;
                Bitmap bitmap = mNormalBitmap;
                int index = i - 1;
                if (index < getCurrentPosition()) {
                    left = (i - 1) * (mNormalBitmapWidth + mIndicatorPadding);
                    top = getMeasuredHeight() / 2 - mNormalBitmapHeight / 2;
                } else if (index == getCurrentPosition()) {
                    left = (i - 1) * (mNormalBitmapWidth + mIndicatorPadding);
                    top = getMeasuredHeight() / 2 - mCheckedBitmapHeight / 2;
                    bitmap = mCheckedBitmap;
                } else {
                    left = (i - 1) * mIndicatorPadding + (i - 2) * mNormalBitmapWidth + mCheckedBitmapWidth;
                    top = getMeasuredHeight() / 2 - mNormalBitmapHeight / 2;
                }
                drawIcon(canvas, left, top, bitmap);
            }
        }
    }

    private void drawIcon(Canvas canvas, int left, int top, Bitmap icon) {
        if (icon == null) {
            return;
        }
        canvas.drawBitmap(icon, left, top, null);
    }

    private void initIconSize() {
        if (mCheckedBitmap != null) {
            if (mIndicatorSize != null) {
                if (mCheckedBitmap.isMutable()&& checkCanResize) {
                    mCheckedBitmap.setWidth(mIndicatorSize.checkedWidth);
                    mCheckedBitmap.setHeight(mIndicatorSize.checkedHeight);
                } else {
                    int width = mCheckedBitmap.getWidth();
                    int height = mCheckedBitmap.getHeight();
                    float scaleWidth = ((float) (mIndicatorSize.checkedWidth) / width);
                    float scaleHeight = ((float) (mIndicatorSize.checkedHeight) / height);
                    Matrix matrix = new Matrix();
                    matrix.postScale(scaleWidth, scaleHeight);
                    mCheckedBitmap = Bitmap.createBitmap(mCheckedBitmap, 0, 0, width, height, matrix, true);
                }
            }
            mCheckedBitmapWidth = mCheckedBitmap.getWidth();
            mCheckedBitmapHeight = mCheckedBitmap.getHeight();
        }

        if (mNormalBitmap != null) {
            if (mIndicatorSize != null) {
                if (mNormalBitmap.isMutable()&& normalCanResize) {
                    mNormalBitmap.setWidth(mIndicatorSize.normalWidth);
                    mNormalBitmap.setHeight(mIndicatorSize.normalHeight);
                } else {
                    int width = mNormalBitmap.getWidth();
                    int height = mNormalBitmap.getHeight();
                    float scaleWidth = ((float) (mIndicatorSize.normalWidth) / mNormalBitmap.getWidth());
                    float scaleHeight = ((float) (mIndicatorSize.normalHeight) / mNormalBitmap.getHeight());
                    Matrix matrix = new Matrix();
                    matrix.postScale(scaleWidth, scaleHeight);
                    mNormalBitmap = Bitmap.createBitmap(mNormalBitmap, 0, 0, width, height, matrix, true);
                }
            }
            mNormalBitmapWidth = mNormalBitmap.getWidth();
            mNormalBitmapHeight = mNormalBitmap.getHeight();
        }
    }

    public DrawableIndicator setIndicatorDrawable(@DrawableRes int normalDrawable, @DrawableRes int checkedDrawable) {
        mNormalBitmap = mCheckedBitmap = BitmapFactory.decodeResource(getResources(), normalDrawable);
        mCheckedBitmap = BitmapFactory.decodeResource(getResources(), checkedDrawable);
        if (mNormalBitmap == null) {
            mNormalBitmap = getBitmapFromVectorDrawable(getContext(), normalDrawable);
            normalCanResize =false;
        }
        if (mCheckedBitmap == null) {
            mCheckedBitmap = getBitmapFromVectorDrawable(getContext(), checkedDrawable);
            checkCanResize =false;
        }
        initIconSize();
        postInvalidate();
        return this;
    }

    public DrawableIndicator setIndicatorSize(int normalWidth, int normalHeight, int checkedWidth, int checkedHeight) {
        this.mIndicatorSize = new IndicatorSize(normalWidth, normalHeight, checkedWidth, checkedHeight);
        initIconSize();
        postInvalidate();
        return this;
    }

    public DrawableIndicator setIndicatorGap(int padding) {
        if (padding >= 0) {
            mIndicatorPadding = padding;
            postInvalidate();
        }
        return this;
    }

    public static Bitmap getBitmapFromVectorDrawable(Context context, int drawableId) {
        Drawable drawable = AppCompatDrawableManager.get().getDrawable(context, drawableId);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            drawable = (DrawableCompat.wrap(drawable)).mutate();
        }
        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }

    static class IndicatorSize {
        int normalWidth;
        int checkedWidth;
        int normalHeight;
        int checkedHeight;

        public IndicatorSize(int normalWidth, int normalHeight, int checkedWidth, int checkedHeight) {
            this.normalWidth = normalWidth;
            this.checkedWidth = checkedWidth;
            this.normalHeight = normalHeight;
            this.checkedHeight = checkedHeight;
        }
    }
}

File: bannerview/src/main/java/com/zhpan/bannerview/indicator/DrawableIndicator.kt
package com.zhpan.bannerview.indicator

import android.content.Context
import android.graphics.*
import android.os.Build
import android.util.AttributeSet
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import com.zhpan.indicator.base.BaseIndicatorView

/**
 * @ author : zhouweibin
 * @ time: 2019/12/18 17:04.
 * @ desc: 选中与未选中的图片长宽可能不一样
 */
class DrawableIndicator @JvmOverloads constructor(context: Context?, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : BaseIndicatorView(context!!, attrs, defStyleAttr) {
    // 选中时的Bitmap
    private var mCheckedBitmap: Bitmap? = null

    // 未选中时的Bitmap
    private var mNormalBitmap: Bitmap? = null

    // 图片之间的间距
    private var mIndicatorPadding = 0

    // 选中图片的宽度
    private var mCheckedBitmapWidth = 0

    // 选中图片的高度
    private var mCheckedBitmapHeight = 0

    //未选中图片的宽高
    private var mNormalBitmapWidth = 0
    private var mNormalBitmapHeight = 0
    private var mIndicatorSize: IndicatorSize? = null
    private var normalCanResize = true
    private var checkCanResize = true
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val maxHeight = mCheckedBitmapHeight.coerceAtLeast(mNormalBitmapHeight)
        val realWidth = mCheckedBitmapWidth + (mNormalBitmapWidth + mIndicatorPadding) * (getPageSize() - 1)
        setMeasuredDimension(realWidth, maxHeight)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (getPageSize() > 1 && mCheckedBitmap != null && mNormalBitmap != null) {
            for (i in 1 until getPageSize() + 1) {
                var left: Int
                var top: Int
                var bitmap = mNormalBitmap
                val index = i - 1
                when {
                    index < getCurrentPosition() -> {
                        left = (i - 1) * (mNormalBitmapWidth + mIndicatorPadding)
                        top = measuredHeight / 2 - mNormalBitmapHeight / 2
                    }
                    index == getCurrentPosition() -> {
                        left = (i - 1) * (mNormalBitmapWidth + mIndicatorPadding)
                        top = measuredHeight / 2 - mCheckedBitmapHeight / 2
                        bitmap = mCheckedBitmap
                    }
                    else -> {
                        left = (i - 1) * mIndicatorPadding + (i - 2) * mNormalBitmapWidth + mCheckedBitmapWidth
                        top = measuredHeight / 2 - mNormalBitmapHeight / 2
                    }
                }
                drawIcon(canvas, left, top, bitmap)
            }
        }
    }

    override fun rotateCanvas(canvas: Canvas) {

    }

    private fun drawIcon(canvas: Canvas, left: Int, top: Int, icon: Bitmap?) {
        if (icon == null) {
            return
        }
        canvas.drawBitmap(icon, left.toFloat(), top.toFloat(), null)
    }

    private fun initIconSize() {
        mCheckedBitmap?.let {
            mIndicatorSize?.let {
                if (mCheckedBitmap!!.isMutable && checkCanResize) {
                    mCheckedBitmap!!.width = mIndicatorSize!!.checkedWidth
                    mCheckedBitmap!!.height = mIndicatorSize!!.checkedHeight
                } else {
                    val width = mCheckedBitmap!!.width
                    val height = mCheckedBitmap!!.height
                    val scaleWidth = mIndicatorSize!!.checkedWidth.toFloat() / width
                    val scaleHeight = mIndicatorSize!!.checkedHeight.toFloat() / height
                    val matrix = Matrix()
                    matrix.postScale(scaleWidth, scaleHeight)
                    mCheckedBitmap = Bitmap.createBitmap(mCheckedBitmap!!, 0, 0, width, height, matrix, true)
                }
            }
            mCheckedBitmapWidth = mCheckedBitmap!!.width
            mCheckedBitmapHeight = mCheckedBitmap!!.height
        }
        mNormalBitmap?.let {
            mIndicatorSize?.let {
                if (mNormalBitmap!!.isMutable && normalCanResize) {
                    mNormalBitmap!!.width = mIndicatorSize!!.normalWidth
                    mNormalBitmap!!.height = mIndicatorSize!!.normalHeight
                } else {
                    val width = mNormalBitmap!!.width
                    val height = mNormalBitmap!!.height
                    val scaleWidth = mIndicatorSize!!.normalWidth.toFloat() / mNormalBitmap!!.width
                    val scaleHeight = mIndicatorSize!!.normalHeight.toFloat() / mNormalBitmap!!.height
                    val matrix = Matrix()
                    matrix.postScale(scaleWidth, scaleHeight)
                    mNormalBitmap = Bitmap.createBitmap(mNormalBitmap!!, 0, 0, width, height, matrix, true)
                }
            }
            mNormalBitmapWidth = mNormalBitmap!!.width
            mNormalBitmapHeight = mNormalBitmap!!.height
        }
    }

    fun setIndicatorDrawable(@DrawableRes normalDrawable: Int, @DrawableRes checkedDrawable: Int): DrawableIndicator {
        mNormalBitmap = BitmapFactory.decodeResource(resources, normalDrawable)
        mCheckedBitmap = BitmapFactory.decodeResource(resources, checkedDrawable)
        if (mNormalBitmap == null) {
            mNormalBitmap = getBitmapFromVectorDrawable(context, normalDrawable)
            normalCanResize = false
        }
        if (mCheckedBitmap == null) {
            mCheckedBitmap = getBitmapFromVectorDrawable(context, checkedDrawable)
            checkCanResize = false
        }
        initIconSize()
        postInvalidate()
        return this
    }

    fun setIndicatorSize(normalWidth: Int, normalHeight: Int, checkedWidth: Int, checkedHeight: Int): DrawableIndicator {
        mIndicatorSize = IndicatorSize(normalWidth, normalHeight, checkedWidth, checkedHeight)
        initIconSize()
        postInvalidate()
        return this
    }

    fun setIndicatorGap(padding: Int): DrawableIndicator {
        if (padding >= 0) {
            mIndicatorPadding = padding
            postInvalidate()
        }
        return this
    }

    internal class IndicatorSize(var normalWidth: Int, var normalHeight: Int, var checkedWidth: Int, var checkedHeight: Int)

    private fun getBitmapFromVectorDrawable(context: Context, drawableId: Int): Bitmap? {
        var drawable = ContextCompat.getDrawable(context, drawableId)
        drawable?.let {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                drawable = DrawableCompat.wrap(drawable!!).mutate()
            }
            val bitmap = Bitmap.createBitmap(drawable!!.intrinsicWidth,
                    drawable!!.intrinsicHeight, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            drawable!!.setBounds(0, 0, canvas.width, canvas.height)
            drawable!!.draw(canvas)
            return bitmap
        }
        return null
    }
}